nrpy/equations/nrpyelliptic/ConformallyFlat_SourceTerms.py

```diff
@@ -0,0 +1,320 @@
+"""
+Generating source terms of Hamiltonian constraint equation using a reference-metric formalism.
+
+Author: Thiago Assumpção
+        assumpcaothiago **at** gmail **dot* com
+
+License: BSD 2-Clause
+"""
+
+# Step Import needed modules:
+from typing import cast, List, Tuple, Union
+import sympy as sp  # For symbolic computations
+import nrpy.indexedexp as ixp  # NRPy+: Symbolic indexed expression (e.g., tensors, vectors, etc.) support
+import nrpy.reference_metric as refmetric  # NRPy+: Reference metric support
+from nrpy.equations.nrpyelliptic.CommonParams import (
+    bare_mass_0,
+    bare_mass_1,
+    zPunc,
+    P0U,
+    P1U,
+    S0U,
+    S1U,
+)
+
+
+def compute_psi_background_and_ADD_times_AUU(
+    CoordSystem: str,
+) -> Tuple[sp.Expr, sp.Expr]:
+    """
+    Compute conformal factor and contraction of conformal extrinsic curvature.
+
+    :param CoordSystem: The coordinate system being used.
+
+    :return: psi_background, ADD_times_AUU
+    """
+    # Step 1: Set up the reference metric
+    rfm = refmetric.reference_metric[CoordSystem]
+
+    # Step 2: Define and Cartesian coordinates
+    xxCart = [rfm.Cartx, rfm.Carty, rfm.Cartz]
+
+    # Step 3: Set relative positions of each puncture
+
+    # Step 3.a: First copy array with Cartesian coordinates
+    x0U = xxCart[:]
+    x1U = xxCart[:]
+    # Step 3.b: Then shift z coordinates
+    x0U[2] -= zPunc
+    x1U[2] += zPunc
+
+    # Step 4: Compute psi_background
+    psi_background_cart = psi_background_cartesian(bare_mass_0, x0U, bare_mass_1, x1U)
+    psi_background = replace_cart_coord_by_xx(rfm, psi_background_cart)
+
+    # Step 5: Define Bowen-York vector field V_i
+    VU_cart = VU_cart_two_punctures(x0U, P0U, S0U, x1U, P1U, S1U)
+
+    # Step 6: Define extrinsic curvature
+    ADD_conf_cart = ADD_conf_cartesian(rfm, VU_cart)
+
+    # Step 7: Compute the scalar ADD*AUU in Cartesian coordinates
+    ADD_times_AUU_conf_cart = ADD_times_AUU_conf_cartesian(ADD_conf_cart)
+
+    # Step 8: Compute ADD_times_AUU in xx coordinates
+    ADD_times_AUU = replace_cart_coord_by_xx(rfm, ADD_times_AUU_conf_cart)
+
+    return psi_background, ADD_times_AUU
+
+
+def replace_cart_coord_by_xx(rfm: refmetric.ReferenceMetric, expr: sp.Expr) -> sp.Expr:
```

Comment:
I'm not a fan of this approach. What should happen is, _in the C code_, xx_to_Cart() should be called. Doing it this way could result in a huge ballooning of expressions.

Reply:
Comments?

Reply:
I understand that calling xx_to_Cart() is more desirable than making this replacement at the codegen level. On the other hand, I used this exact same approach to generate the code that was ultimately used in the NRPyElliptic announcement paper (https://arxiv.org/abs/2111.02424). I never experience any issues with having very large expressions or slow codegen.

I would be happy to change this. On the other hand, I wanted to have this version as close as possible to the first NRPyElliptic version, to facilitate future comparisons. We could always come back and change this, once the whole C project in constructed.

Reply:
Okay. Please add a # FIXME : replace with xx_to_Cart() in C code, and I'll approve the PR.

Reply:
Done.